src/rebase.rs

```diff
@@ -0,0 +1,487 @@
+        // invocation and we need to offset idx + 1 for the general "are we
+        // done with the loop" query.
+        match self.rebase.current_operation_index() {
+            None => Some(self.rebase.next()),
+            Some(idx) if idx + 1 < self.operation_count => Some(self.rebase.next()),
+            _ => None
+        }
+    }
+}
+
+impl<'repo> Binding for Rebase<'repo> {
+    type Raw = *mut raw::git_rebase;
+    unsafe fn from_raw(raw: *mut raw::git_rebase) -> Rebase<'repo> {
+        Rebase {
+            raw: raw,
+            _marker: marker::PhantomData,
+        }
+    }
+    fn raw(&self) -> *mut raw::git_rebase {
+        self.raw
+    }
+}
+
+impl<'repo> Debug for Rebase<'repo> {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        f.debug_struct("Rebase")
+            .field("operation_count", &self.operation_count())
+            .field("current_operation_index", &self.current_operation_index())
+            .finish()
+    }
+}
+
+impl<'repo> Drop for Rebase<'repo> {
+    fn drop(&mut self) {
+        unsafe { raw::git_rebase_free(self.raw) }
+    }
+}
+
+impl<'rebase, 'repo: 'rebase> Binding for RebaseOperation<'rebase, 'repo> {
+    type Raw = *mut raw::git_rebase_operation;
+    unsafe fn from_raw(raw: *mut raw::git_rebase_operation) -> RebaseOperation<'rebase, 'repo> {
+        RebaseOperation {
+            raw: raw,
+            _marker: marker::PhantomData,
+        }
+    }
+    fn raw(&self) -> *mut raw::git_rebase_operation {
+        self.raw
+    }
+}
+
+impl<'rebase, 'repo: 'rebase> Debug for RebaseOperation<'rebase, 'repo> {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        f.debug_struct("RebaseOperation")
+            .field("id", &self.id())
+            .field("kind", &self.kind())
+            .finish()
+    }
+}
+
+impl<'a> RebaseOptions<'a> {
+    /// Initiate default set of rebase options.
+    pub fn new() -> RebaseOptions<'a> {
+        RebaseOptions {
+            version: raw::GIT_REBASE_OPTIONS_VERSION as usize,
+            quiet: false,
+            in_memory: false,
+            rewrite_notes_ref: None,
+            merge_options: None,
+            checkout_builder: None,
+        }
+    }
+
+    /// this will instruct other clients working
+    /// on this `rebase` that you want a quiet rebase experience.
+    /// This is provided for interoperability between Git tools
+    pub fn quiet(&mut self, quiet: bool) -> &mut Self {
+        self.quiet = quiet;
+        self
+    }
+
+    /// Perform an in-memory rebase, will not updated the repository to be in a rebasing-state
+    /// or modify the working directory.
+    pub fn in_memory(&mut self, in_memory: bool) -> &mut Self {
+        self.in_memory = in_memory;
+        self
+    }
+
+    /// The name of the notes reference used to rewrite notes
+    /// for rebased commits when finishing the rebase
+    pub fn rewrite_notes_ref(&mut self, rewrite_notes_ref: Option<CString>) -> &mut Self {
```

Comment:
I think idiomatically currently this library uses `fn rewrite_notes_ref<T: IntoCString>(&mut self, ref: T) -> ...` to avoid the `Option<CString>`

Reply:
Hmm `IntoCString` doesn't support returning an `Option<CString>`, just a `Result`, and it's not implemented on `Option<IntoCString>`. I could do the Into dance:

```rust
    pub fn rewrite_notes_ref<T, U>(&mut self, rewrite_notes_ref: T) -> &mut Self
    where
        T: Into<Option<U>>,
        U: IntoCString,
    {
        self.rewrite_notes_ref = rewrite_notes_ref.into().map(|s| s.into_c_string().unwrap());
        self
    }
```

Possibly transforming the return value into a result, but I think that the chance that a ref isn't a valid CString _probably_ requires pretty crazy input. I'll put that up as the change, but lmk if you'd prefer some other approach.